From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Christopher Drelich <cd@cdrakka.com>
Date: Wed, 23 May 2018 22:50:38 -0400
Subject: [PATCH] restartsig: restart dwm via MOD+CTRL+SHIFT+Q or kill -HUP

Adapted from dwm-restartsig-20180523-6.2.diff for dwm 6.5.
Modified quit() to restart if it receives arg .i = 1.
Signal handlers added for SIGHUP (restart) and SIGTERM (quit).
---
diff --git a/config.def.h b/config.def.h
--- a/config.def.h
+++ b/config.def.h
@@ -94,6 +94,7 @@
 	TAGKEYS(                        XK_8,                      7)
 	TAGKEYS(                        XK_9,                      8)
 	{ MODKEY|ShiftMask,             XK_q,      quit,           {0} },
+	{ MODKEY|ControlMask|ShiftMask, XK_q,      quit,           {1} },
 };
 
 /* button definitions */
diff --git a/dwm.c b/dwm.c
--- a/dwm.c
+++ b/dwm.c
@@ -207,5 +207,7 @@ static void seturgent(Client *c, int urg);
 static void showhide(Client *c);
 static void spawn(const Arg *arg);
 static void tag(const Arg *arg);
+static void sighup(int unused);
+static void sigterm(int unused);
 static void tagmon(const Arg *arg);
 static void togglebar(const Arg *arg);
@@ -261,6 +263,7 @@ static void (*handler[LASTEvent]) (XEvent *) = {
 };
 static Atom wmatom[WMLast], netatom[NetLast];
 static int running = 1;
+static int restart = 0;
 static Cur *cursor[CurLast];
 static Clr **scheme;
 static Display *dpy;
@@ -1259,6 +1262,7 @@ quit(const Arg *arg)
 void
 quit(const Arg *arg)
 {
+	if(arg->i) restart = 1;
 	running = 0;
 }
 
@@ -1554,6 +1558,9 @@ setup(void)
 	sigaction(SIGCHLD, &sa, NULL);
 
 	/* clean up any zombies (inherited from .xinitrc etc) immediately */
+	signal(SIGHUP, sighup);
+	signal(SIGTERM, sigterm);
+
 	while (waitpid(-1, NULL, WNOHANG) > 0);
 
 	/* init screen */
@@ -1648,6 +1655,20 @@ spawn(const Arg *arg)
 }
 
 void
+sighup(int unused)
+{
+	Arg a = {.i = 1};
+	quit(&a);
+}
+
+void
+sigterm(int unused)
+{
+	Arg a = {.i = 0};
+	quit(&a);
+}
+
+void
 tag(const Arg *arg)
 {
 	if (selmon->sel && arg->ui & TAGMASK) {
@@ -2162,6 +2183,7 @@ main(int argc, char *argv[])
 	scan();
 	run();
 	cleanup();
+	if(restart) execvp(argv[0], argv);
 	XCloseDisplay(dpy);
 	return EXIT_SUCCESS;
 }
